075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,435 @@
+3. On the **Nth** reconciliation (rebalancing is done):
+   * the `podSet()` does nothing because no new brokers are added this time.
+   * the `KafkaAutorebalancingReconciler().reconcile()` detects that the `KafkaRebalance` custom resource status reports the `Ready` state (so the rebalancing is completed). The end of rebalancing is reflected into the `Kafka.status.autoRebalance.state` field and the `KafkaRebalance` is deleted.
+
+Because the rebalancing runs across several reconciliations, different scenarios could happen even taking into account operator restarts:
+
+* the `KafkaCluster.addedNodes()` list is empty:
+  * the `Kafka.status.autoRebalance.state` is at `RebalancingOnScaleUp`. It's the normal flow, so just checking the rebalancing state.
+  * the `Kafka.status.autoRebalance.state` doesn't exist or it's `Ready`. It means no scale up requested. Nothing to do in terms of rebalancing.
+* the `KafkaCluster.addNodes()` list is not empty:
+  * the `Kafka.status.autoRebalance.state` doesn't exist or it's `Ready`. It's the normal flow, a new rebalance has to start.
+  * the `Kafka.status.autoRebalance.state` is at `RebalancingOnScaleUp`. The newly added brokers' IDs are compared with the one in the `Kafka.status.autoRebalance.modes.brokers` field. If they are equal, nothing to do (even if this situation should not be possible). If they are different, the user could have asked for another scale up right after a previous one. The operator has to stop the current rebalancing, by applying the `strimzi.io/rebalance: stop` annotation on the corresponding `KafkaRebalance` custom resource, and start a new rebalancing, by applying the `strimzi.io/rebalance: refresh` annotation, with a consistent list of brokers defined by the multiple requested scales up.
+
+During the above flow, errors can happen after the scaling up:
+
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
+
+In both scenarios, the reconciliation ends with the `KafkaRebalance` custom resource in the `NotReady` state which is reflected into the `Kafka.status.autoRebalance.state` field.
+
+On the next reconciliation, the scaling up is done already, but the rebalancing cannot run because of the `KafkaRebalance` custom resource in the `NotReady` state.
+The user can fix it by deleting the `KafkaRebalance` custom resource so that on the next reconciliation, the operator will be able to retry an auto-rebalancing, by recreating a new `KafkaRebalance` custom resource getting the added brokers from the `status.autoRebalancing.brokers` for the `add-brokers` mode.
+Another approach the user can take is removing the `spec.cruiseControl.autoRebalance` (other than deleting the `KafkaRebalance` custom resource) in order to run a manual rebalancing later on.
```

Comment:
This one seems super weird. But I'm not sure I have a better idea how the user can tell the operator to just give up. Should we maybe approach it differently? When the scale rebalance fails, you reset the add brokers in the status to not continue the rebalance for these brokers and the user completes it manually?